runtime/runtime-core/common/test/aws/smithy/kotlin/runtime/retries.impl/StandardRetryPolicyTest.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.retries.impl
+
+import aws.smithy.kotlin.runtime.*
+import aws.smithy.kotlin.runtime.retries.RetryDirective
+import aws.smithy.kotlin.runtime.retries.RetryErrorType
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+class StandardRetryPolicyTest {
+    private fun test(ex: Throwable): RetryDirective = StandardRetryPolicy().evaluate(Result.failure(ex))
+
+    @Test
+    fun testSuccess() {
+        val result = StandardRetryPolicy().evaluate(Result.success("Literally any value will work!"))
+        assertEquals(RetryDirective.TerminateAndSucceed, result)
+    }
+
+    @Test
+    fun testThrottling() {
+        val ex = SdkBaseException()
+        ex.sdkErrorMetadata.attributes[ErrorMetadata.Throttling] = true
+        val result = test(ex)
+        assertEquals(RetryDirective.RetryError(RetryErrorType.Throttling), result)
+    }
+
+    @Test
+    fun testServerSideException() {
+        val ex = ServiceException()
+        ex.sdkErrorMetadata.attributes[ErrorMetadata.Retryable] = true
+        ex.sdkErrorMetadata.attributes[ServiceErrorMetadata.ErrorType] = ServiceException.ErrorType.Server
+        val result = test(ex)
+        assertEquals(RetryDirective.RetryError(RetryErrorType.ServerSide), result)
+    }
+
+    @Test
+    fun testClientSideException() {
+        val ex = ServiceException()
+        ex.sdkErrorMetadata.attributes[ErrorMetadata.Retryable] = true
+        ex.sdkErrorMetadata.attributes[ServiceErrorMetadata.ErrorType] = ServiceException.ErrorType.Client
+        val result = test(ex)
+        assertEquals(RetryDirective.RetryError(RetryErrorType.ClientSide), result)
+    }
+
+    @Test
+    fun testClientException() {
+        val result = test(ClientException())
+        assertEquals(RetryDirective.RetryError(RetryErrorType.ClientSide), result)
```

Comment:
question

should this attempt to retry any `ClientException` or should this terminate and fail? I would assume most client side (if not all) are not actually retryable.

Reply:
That's a good question and I suppose it depends on what else is a `ClientException`. We just decided to make all of the retryer exceptions (e.g., `TimedOutException`) inherit from `ClientException` and those are technically retryable (in that they _could_ succeed with more tries).

Taking a look through the runtime, looks like most other instances are serialization (likely not retryable) or configuration related (possibly retryable if the configuration is coming from somewhere remote).  Any other instances we can think of?

Reply:
No but in general `ClientException` is basically reserved for runtime related exceptions (not to be confused with `ServiceException` where the `ErrorType` is set to `Client`)